src/test/java/com/fasterxml/jackson/databind/deser/TestCustomSubTypeValidator.java

```diff
@@ -0,0 +1,277 @@
+            _name = name;
+        }
+    }
+
+    public static class Lion extends Animal
+    {
+    }
+
+    public static class Warthog extends Animal
+    {
+    }
+
+    public static class Goat extends Animal
+    {
+    }
+
+    public static class Sheep extends Animal
+    {
+    }
+
+    public static class Penguin extends Animal
+    {
+    }
+
+    /*
+    /**********************************************************
+    /* Helper classes (validators)
+    /**********************************************************
+     */
+
+    static class DenyAllValidator
+        extends SubTypeValidator
+    {
+        @Override
+        public void validateSubType(DeserializationContext ctxt, JavaType type, BeanDescription beanDesc)
+            throws JsonMappingException 
+        {
+            ctxt.reportBadTypeDefinition(
+                    beanDesc,
+                    "Illegal type (%s) to deserialize: this validator does not allow deserialization at all!",
+                    type.getRawClass().getName());
+        }
+    }
+
+    static class FriendlyAnimalValidator
+        extends SubTypeValidator
+    {
+        private static Class<?>[] FRIENDLY_ANIMAL_WHITELIST = new Class<?>[] {
+            Animal.class,
+            Goat.class,
+            Sheep.class,
+            Penguin.class
+        };
+
+        @Override
+        public void validateSubType(DeserializationContext ctxt, JavaType type, BeanDescription beanDesc)
+            throws JsonMappingException 
+        {
+            super.validateSubType(ctxt, type, beanDesc);
+
+            if (type.isTypeOrSubTypeOf(Animal.class)) {
+                for (final Class<?> friendlyAnimalSpecies : FRIENDLY_ANIMAL_WHITELIST) {
+                    if (type.hasRawClass(friendlyAnimalSpecies)) {
+                        return;
+                    }
+                }
+
+                ctxt.reportBadTypeDefinition(
+                        beanDesc,
+                        "Illegal type (%s) to deserialize: prevented because this animal species is not known to be friendly",
+                        type.getRawClass().getName());
+            }
+        }
+    }
+
+    /*
+    /**********************************************************
+    /* Factory methods
+    /**********************************************************
+     */
+
+    protected ObjectMapper objectMapper(SubTypeValidator validator)
+    {
+        final DeserializerFactoryConfig config =
+                new DeserializerFactoryConfig()
+                    .withSubTypeValidator(validator);
+
+        final DeserializerFactory factory =
+                new BeanDeserializerFactory(config);
+
+        final JsonMapper mapper =
+                JsonMapper
+                    .builder()
+                    .deserializerFactory(factory)
+                    .build();
+
+        return mapper;
+    }
+
+    /*
+    /**********************************************************
+    /* Unit tests
```

Comment:
I have left the number of assertions at a minimum. Ultimately, I focused on testing whether a specific validator will reject inputs or not. I can certainly add more assertions on the deserialized data?